File: ChordVoicingsCreator/src/main/java/com/campoy/chord/voicing/creator/generators/ChordVoicingGenerator.java
package com.campoy.chord.voicing.creator.generators;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.net.URL;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.Set;
import java.util.SortedSet;
import java.util.StringJoiner;
import java.util.TreeSet;
import java.util.function.Predicate;

import com.campoy.chord.voicing.creator.model.guitar.ChordVoicing;
import com.campoy.chord.voicing.creator.model.guitar.FretAction;
import com.campoy.chord.voicing.creator.model.guitar.GuitarString;
import com.campoy.chord.voicing.creator.model.guitar.KeyAndRoot;
import com.campoy.chord.voicing.creator.model.guitar.Tuning;
import com.campoy.chord.voicing.creator.model.musictheory.ChordConstructor;
import com.campoy.chord.voicing.creator.model.musictheory.Interval;
import com.campoy.chord.voicing.creator.model.musictheory.Note;
import com.campoy.chord.voicing.creator.model.musictheory.NoteAndOctave;
import com.campoy.chord.voicing.creator.model.musictheory.Key;

import lombok.Data;

public class ChordVoicingGenerator {

    /**
     * Generates chord voicings for stringed instruments such as guitar, bass, mandolin, etc, 
     * and filter the chord voicings to find the useful ones 
     * 
     */
    
    
    public static void main(String[] args) throws FileNotFoundException {
    	        
        ChordVoicingGenerator chordVoicingGenerator = new ChordVoicingGenerator();  
        
        Tuning tuning = Tuning.DROP_D_6_STRING_GUITAR;  
        
        int fretSpan = 3; // maximum fret span that the voicings are allowed to have / number of frets that one can comfortably span with their fingers  
        int firstFretToScan = 5;  
        int lastFretToScan = 11;  
        
        // minimum number of semitones between notes of the voicing, used by example to  
        // filter out voicings where the maj7th/min2nd and the root are right next to each other  
        int minimumSemitonesBetweenNotesOfTheVoicing  = 2;  
        
        List<Key> searchedKeys = Arrays.asList(  
                Key.AEOLIAN  
                , Key.HARMONIC_MINOR  
                );  
        
        List<Note> allowedKeyRoots = Arrays.asList(  
                Note.E);  
        
        int minNumberOfDifferentNotes = 3;  
        int maxNumberOfDifferentNotes = 4;  
        int mustContainAtLeastThisManyOpenStrings = 0;  
        int mustContainAtMostThisManyOpenStrings = 2;  
        int minimumNumberOfMutedLowestStrings = 0;  
        int minimumNumberOfMutedHighestStrings = 0;  
        
        boolean forbidTheSameNoteOnTheSameOctaveOnDifferentStrings = true;  
        
        NoteAndOctave lowestNoteAllowed = new NoteAndOctave(Note.E, 3);  
        NoteAndOctave highestNoteAllowed = new NoteAndOctave(Note.E, 6);  
        
        // avoid showing chord voicings that have neither a 5th nor a third (to not show  
        // that a E minor could be considered a B add11 b13 that doesn't have a  
        // third or a fifth). That makes things more readable  
        boolean hideChordVoicingsThatDoNotHaveAThirdOrAFifth = true;  
        
        // ------
        
        Map<Key, Map<Note, List<Note>>> keyToMapOfRootsToNotesOfKey = 
                computeMapOfKeyToMapOfRootsToNotesOfKey(searchedKeys, allowedKeyRoots);
                    
        List<Predicate<ChordVoicing>> filters = new ArrayList<>();
        filters.add(voicing -> {
            return voicing.smallestDistanceBetweenVoices() >= minimumSemitonesBetweenNotesOfTheVoicing;
        });
        filters.add(voicing -> {
            return voicing.isCompatibleWithAnyOfTheseKeys(keyToMapOfRootsToNotesOfKey);
        });
        filters.add(voicing -> {
            return voicing.getRepresentedChord().getNotes().size() <= maxNumberOfDifferentNotes;
        });
        filters.add(voicing -> {
            return voicing.getRepresentedChord().getNotes().size() >= minNumberOfDifferentNotes;
        });
        filters.add(voicing -> {
        	
        	return forbidTheSameNoteOnTheSameOctaveOnDifferentStrings 
        			&& !voicing.getHasSeveralTimesTheSameNoteOnTheSameOctave();

        });
        filters.add(voicing -> {
        	int numberOfOpenStrings = 0;
        	for(Entry<GuitarString, FretAction> entry : voicing.getFrettings().entrySet()) {
        		if(entry.getValue().isOpen()) {
                	numberOfOpenStrings++;
                }
        	}
            return (numberOfOpenStrings >= mustContainAtLeastThisManyOpenStrings);
        });
        filters.add(voicing -> {
            int numberOfOpenStrings = 0;
            for(Entry<GuitarString, FretAction> entry : voicing.getFrettings().entrySet()) {
                if(entry.getValue().isOpen()) {
                    numberOfOpenStrings++;
                }
            }
            return (numberOfOpenStrings <= mustContainAtMostThisManyOpenStrings);
        });
        filters.add(voicing -> {
            
            boolean passed = true;
            for( int stringNumber = 0; 
                    stringNumber < minimumNumberOfMutedLowestStrings ;
                    stringNumber++) {
                
                passed &= FretAction.MUTE.equals(voicing.getFrettings().get(new GuitarString(stringNumber)));
            }
            return passed;
            
        });
        filters.add(voicing -> {
            
            boolean passed = true;
            for( int stringNumber = tuning.getNumberOfStrings() - 1; 
                    stringNumber + minimumNumberOfMutedHighestStrings >= tuning.getNumberOfStrings() ;
                    stringNumber--) {
                
                passed &= FretAction.MUTE.equals(voicing.getFrettings().get(new GuitarString(stringNumber)));
            }
            return passed;
            
        });
        filters.add(voicing -> {
            return highestNoteAllowed == null 
                    || voicing.getHighestNote(tuning).isLowerThan(highestNoteAllowed);
            
        });
        filters.add(voicing -> {
            return lowestNoteAllowed == null 
                    || voicing.getLowestNote(tuning).isHigherThan(lowestNoteAllowed);
        
        });
        filters.add(voicing -> {
            return !hideChordVoicingsThatDoNotHaveAThirdOrAFifth 
                    || voicing.doAnyOfTheRepresentationsHaveAThirdOrAFifth(tuning);
        });
        
        
        
//        Path filePath = Paths.get("C:/Users/Mickael/eclipse-workspace/PancakeMusicProjects/Voices.txt");
        Path filePath = Paths.get("./Voices.txt");
        System.out.println("Printing to " + filePath.toAbsolutePath());
        File targetFile = filePath.toFile();
        FileOutputStream o = new FileOutputStream(targetFile);
        BufferedOutputStream bos = new BufferedOutputStream(o);
        PrintStream out = new PrintStream(bos);
        
        try {
            
        
            chordVoicingGenerator.generateAllVoicings(
        		tuning, 
        		fretSpan,
        		firstFretToScan,
        		lastFretToScan, 
        		filters,
        		out,
        		keyToMapOfRootsToNotesOfKey,
        		hideChordVoicingsThatDoNotHaveAThirdOrAFifth);
        }
        finally {
            
            out.close();
        }
        
        System.out.println("stream closed");
                
    }

    private static Map<Key, Map<Note, List<Note>>> computeMapOfKeyToMapOfRootsToNotesOfKey(
            List<Key> searchedKeys, List<Note> allowedRoots) {
        Map<Key, Map<Note, List<Note>>> keyToMapOfRootsToNotesOfKey = new HashMap<>();
        
        for(Key searchedKey : searchedKeys ) {
            Map<Note, List<Note>> rootsToNotesOfKey = new HashMap<>();
            for(Note root : Note.values()) {
                
                if( allowedRoots == null 
                        || allowedRoots.isEmpty()
                        || allowedRoots.contains(root)) {
                
                    List<Note> notesInKeyStartingOnRoot = new ArrayList<Note>();
                    List<Integer> semitonesFromRootList = searchedKey.getSemitonesFromRootList();
                    for (Integer semitonesFromRoot : semitonesFromRootList) {
                        notesInKeyStartingOnRoot.add(root.up(semitonesFromRoot));
                    }
                    rootsToNotesOfKey.put(root, notesInKeyStartingOnRoot);
                }
            }
            keyToMapOfRootsToNotesOfKey.put(searchedKey, rootsToNotesOfKey);
        }
        return keyToMapOfRootsToNotesOfKey;
    }

    private void generateAllVoicings(
            Tuning tuning,
            int fretSpan,
            int firstFretToScan,
            int lastFretToScan,
            List<Predicate<ChordVoicing>> filters,
            PrintStream output, Map<Key, Map<Note, List<Note>>> keyToMapOfRootsToNotesOfKey,
            boolean hideChordVoicingsThatDoNotHaveAThirdOrAFifth) {
        
        System.out.println("Generating voicings for the tuning " + tuning );
        
        List<StartAndEnd> startEnds = new ArrayList<>();
        
        int fretStart = firstFretToScan;
        int fretEnd;
        while( (fretEnd = fretStart + fretSpan - 1) <= lastFretToScan ) {
        	
        	startEnds.add(new StartAndEnd(fretStart, fretEnd));
            fretStart++;
        }
        
        startEnds.stream().forEach((startEnd) -> {
        	
            Set<ChordVoicing> voicings = generateVoicings(
                    tuning , startEnd.getStart(), startEnd.getEnd(), filters, keyToMapOfRootsToNotesOfKey); 
            
            System.out.println(
                    "Voicings between frets " 
                    + startEnd.getStart() 
                    + " and " + startEnd.getEnd() 
                    + " generated succesfully");
            
            for(ChordVoicing voicing : voicings) {
                                
                output.println("____________ " + voicing.noteRepresentation(tuning) + "  ");
                
            	for( String fullRepresentation : voicing.fullRepresentations(tuning, hideChordVoicingsThatDoNotHaveAThirdOrAFifth)) {
            	    output.println(
                            "_________ " 
                            + fullRepresentation
                            + ":  ");
            	}
            	output.println(voicing.toString() + "  ");
            	output.println(voicing.notesOnStrings() + "  ");
            	StringJoiner sj = new StringJoiner(", ");
            	for(KeyAndRoot compatibleKeyAndRoot : voicing.getCompatibleKeysAndRoots()) {
            	    sj.add(compatibleKeyAndRoot.getRoot() 
            	            + " " 
            	            + compatibleKeyAndRoot.getKey());
            	}
                output.println("Diatonic to " + sj.toString() + "  ");

            	output.println("__________________________  ");
            }
                    
        });
        
        output.println("All voicings generated succesfully" );
        
    }
    
    Set<ChordVoicing> generateVoicings(
            Tuning tuning, 
            Integer lowestFretToScan, 
            Integer highestFretToScan,
            List<Predicate<ChordVoicing>> filters,
            Map<Key, Map<Note, List<Note>>> keyToMapOfRootsToNotesOfKey){
                
        if(lowestFretToScan == null
                || lowestFretToScan == 0) {
            lowestFretToScan = 1; // open is not a something that can be fretted
        }
        if(highestFretToScan == null) {
            highestFretToScan = 5;
        }
        
        Map<Integer, Map<Integer, NoteAndOctave>> stringNumberToFretNumberToNote
            = new HashMap<>();
        
        for( Entry<GuitarString, NoteAndOctave> stringTuning 
                : tuning.getStringNotes().entrySet())
        {
            NoteAndOctave stringTuningNote = stringTuning.getValue();
            int stringNumber = stringTuning.getKey().getPosition();
            if(stringNumberToFretNumberToNote.get(stringNumber) == null) {
                stringNumberToFretNumberToNote.put(stringNumber, new HashMap<>());
            }
            Map<Integer, NoteAndOctave> fretNumberToNote = stringNumberToFretNumberToNote.get(stringNumber);
            for( int currentFret  = lowestFretToScan; currentFret < highestFretToScan ; currentFret++ ) {
                NoteAndOctave currentNote = stringTuningNote.up(currentFret);
                fretNumberToNote.put(currentFret, currentNote);
            }
        }
        
        List<FretAction> possibleFretActions = new ArrayList<>();
        possibleFretActions.add(FretAction.MUTE);
        possibleFretActions.add(FretAction.OPEN);
        for( int currentFret  = lowestFretToScan; currentFret <= highestFretToScan ; currentFret++ ) {
            possibleFretActions.add(FretAction.hold(currentFret));
        }
        
        return generateVoicingsFromFretActions(possibleFretActions, tuning, filters, keyToMapOfRootsToNotesOfKey);
        
    }
    
    Set<ChordVoicing> generateVoicingsFromFretActions(
            List<FretAction> possibleFretActions,
            Tuning tuning,
            List<Predicate<ChordVoicing>> filters,
            Map<Key, Map<Note, List<Note>>> keyToMapOfRootsToNotesOfKey) {
        
        return iterateFrets(possibleFretActions, tuning, 0, new ArrayList<>(), filters, keyToMapOfRootsToNotesOfKey);
    }
    
    Set<ChordVoicing> iterateFrets(
            List<FretAction> possibleFretActions,
            Tuning tuning,
            int currentString, 
            List<FretAction> previousIterations,
            List<Predicate<ChordVoicing>> filters,
            Map<Key, Map<Note, List<Note>>> keyToMapOfRootsToNotesOfKey) {
        
        Set<ChordVoicing> voicingsConstructed = new HashSet<>();
        
        int stringsToIterate = tuning.getNumberOfStrings();
        
        for(FretAction fretAction : possibleFretActions) {
            
            List<FretAction> thisIteration = new ArrayList<>();
            thisIteration.addAll(previousIterations);
            thisIteration.add(fretAction);
                        
            if(currentString < stringsToIterate - 1) {
                
                voicingsConstructed.addAll(
                        iterateFrets(
                                possibleFretActions,
                                tuning,
                                currentString + 1,
                                thisIteration,
                                filters,
                                keyToMapOfRootsToNotesOfKey));
            }
            else {
            	            	
            	ChordVoicing chordVoicing = new ChordVoicing(thisIteration);
            	
            	if(chordVoicing.numberOfNonMutedStrings()  > 1 ) {
	            	chordVoicing.postProcessing(tuning, keyToMapOfRootsToNotesOfKey);
					boolean passedAllFilters = true;
					for(Predicate<ChordVoicing> filter : filters) {
					    if(!filter.test(chordVoicing)){
					        passedAllFilters = false;
					        break;
					    }
					}
					if(passedAllFilters) {
		                voicingsConstructed.add(chordVoicing);
					}  
            	}
            }
            
        }
        
        return voicingsConstructed;
    }
    
    @Data
    private class StartAndEnd {
        private final Integer start;
        private final Integer end;
    }
    
}

File: PancakeMusicProjectsTS/src/generators/ChordVoicingGenerator.ts
import { ChordVoicing } from "../model/guitar/ChordVoicing.js";
import { FretAction } from "../model/guitar/FretAction.js";
import { Tuning } from "../model/guitar/Tuning.js";
import { Key } from "../model/musictheory/Key.js";
import { Note } from "../model/musictheory/Note.js";
import { NoteAndOctave } from "../model/musictheory/NoteAndOctave.js";
import { PrintStream } from "../util/PrintStream.js";

console.log("loading ChordVoicingGenerator")

export class StartAndEnd {
    constructor(public start: number, public end: number) { }
}

export class ChordVoicingGenerator {

    public static generateAndDownloadChordVoicings(
        tuning: Tuning,
        fretSpan: number,
        firstFretToScan: number,
        lastFretToScan: number,
        minimumSemitonesBetweenNotesOfTheVoicing: number,
        searchedKeys: Key[],
        allowedKeyRoots: Note[],
        minNumberOfDifferentNotes: number,
        maxNumberOfDifferentNotes: number,
        mustContainAtLeastThisManyOpenStrings: number,
        mustContainAtMostThisManyOpenStrings: number,
        minimumNumberOfMutedLowestStrings: number,
        minimumNumberOfMutedHighestStrings: number,
        forbidTheSameNoteOnTheSameOctaveOnDifferentStrings: boolean,
        lowestNoteAllowed: NoteAndOctave,
        highestNoteAllowed: NoteAndOctave,
        hideChordVoicingsThatDoNotHaveAThirdOrAFifth: boolean,
        hideChordVoicingsWithb5: boolean,
        hideChordVoicingsWithsharp5: boolean,
        hideChordVoicingsWithb7: boolean,
        hideChordVoicingsWithAdd7: boolean,
        hideChordVoicingsWithb9: boolean,
        hideChordVoicingsWithAdd9: boolean,
        hideChordVoicingsWithb11: boolean,
        hideChordVoicingsWithAdd11: boolean,
        hideChordVoicingsWithb13: boolean,
        hideChordVoicingsWithAdd13: boolean
    ): void {

        // TODO hook the hide chord xxx to the generator
        // make the hide chords look okay
        // make the github pages

        console.log('ChordVoicingGeneratorConfig.tuning:', tuning);
        console.log('ChordVoicingGeneratorConfig.fretSpan:', fretSpan);
        console.log('ChordVoicingGeneratorConfig.firstFretToScan:', firstFretToScan);
        console.log('ChordVoicingGeneratorConfig.lastFretToScan:', lastFretToScan);
        console.log('ChordVoicingGeneratorConfig.minimumSemitonesBetweenNotesOfTheVoicing:', minimumSemitonesBetweenNotesOfTheVoicing);
        console.log('ChordVoicingGeneratorConfig.searchedKeys:', searchedKeys);
        console.log('ChordVoicingGeneratorConfig.allowedKeyRoots:', allowedKeyRoots);
        console.log('ChordVoicingGeneratorConfig.minNumberOfDifferentNotes:', minNumberOfDifferentNotes);
        console.log('ChordVoicingGeneratorConfig.maxNumberOfDifferentNotes:', maxNumberOfDifferentNotes);
        console.log('ChordVoicingGeneratorConfig.mustContainAtLeastThisManyOpenStrings:', mustContainAtLeastThisManyOpenStrings);
        console.log('ChordVoicingGeneratorConfig.mustContainAtMostThisManyOpenStrings:', mustContainAtMostThisManyOpenStrings);
        console.log('ChordVoicingGeneratorConfig.minimumNumberOfMutedLowestStrings:', minimumNumberOfMutedLowestStrings);
        console.log('ChordVoicingGeneratorConfig.minimumNumberOfMutedHighestStrings:', minimumNumberOfMutedHighestStrings);
        console.log('ChordVoicingGeneratorConfig.forbidTheSameNoteOnTheSameOctaveOnDifferentStrings:', forbidTheSameNoteOnTheSameOctaveOnDifferentStrings);
        console.log('ChordVoicingGeneratorConfig.lowestNoteAllowed:', lowestNoteAllowed);
        console.log('ChordVoicingGeneratorConfig.highestNoteAllowed:', highestNoteAllowed);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsThatDoNotHaveAThirdOrAFifth:', hideChordVoicingsThatDoNotHaveAThirdOrAFifth);

        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithb5:', hideChordVoicingsWithb5);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithsharp5:', hideChordVoicingsWithsharp5);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithb7:', hideChordVoicingsWithb7);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithAdd7:', hideChordVoicingsWithAdd7);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithb9:', hideChordVoicingsWithb9);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithAdd9:', hideChordVoicingsWithAdd9);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithb11:', hideChordVoicingsWithb11);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithAdd11:', hideChordVoicingsWithAdd11);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithb13:', hideChordVoicingsWithb13);
        console.log('ChordVoicingGeneratorConfig.hideChordVoicingsWithAdd13:', hideChordVoicingsWithAdd13);

        hideChordVoicingsWithb5
        hideChordVoicingsWithsharp5
        hideChordVoicingsWithb7
        hideChordVoicingsWithAdd7
        hideChordVoicingsWithb9
        hideChordVoicingsWithAdd9
        hideChordVoicingsWithb11
        hideChordVoicingsWithAdd11
        hideChordVoicingsWithb13
        hideChordVoicingsWithAdd13

        console.log('___________________');

        if (firstFretToScan < 1) {
            firstFretToScan = 1;
        }
        if (lastFretToScan < 1) {
            lastFretToScan = 1;
        }

        //default values for testing purposes
        //tuning = Tuning.DROP_D_6_STRING_GUITAR;
        //fretSpan = 3;
        //firstFretToScan = 5;
        //lastFretToScan = 11;
        //minimumSemitonesBetweenNotesOfTheVoicing = 2;
        //searchedKeys= [Key.AEOLIAN, Key.HARMONIC_MINOR];
        //allowedKeyRoots = [Note.E];
        //minNumberOfDifferentNotes = 3;
        //maxNumberOfDifferentNotes = 4;
        //mustContainAtLeastThisManyOpenStrings = 0;
        //mustContainAtMostThisManyOpenStrings = 2;
        //minimumNumberOfMutedLowestStrings = 0;
        //minimumNumberOfMutedHighestStrings = 0;
        //forbidTheSameNoteOnTheSameOctaveOnDifferentStrings = true;
        //lowestNoteAllowed = new NoteAndOctave(Note.E, 3);
        //highestNoteAllowed = new NoteAndOctave(Note.E, 6);
        //hideChordVoicingsThatDoNotHaveAThirdOrAFifth = true;
        //hideChordVoicingsWithb5 = true;
        //hideChordVoicingsWithsharp5 = true;
        //hideChordVoicingsWithb7 = true;
        //hideChordVoicingsWithAdd7 = true;
        //hideChordVoicingsWithb9 = true;
        //hideChordVoicingsWithAdd9 = true;
        //hideChordVoicingsWithb11 = true;
        //hideChordVoicingsWithAdd11 = true;
        //hideChordVoicingsWithb13 = true;
        //hideChordVoicingsWithAdd13 = true;

        const keyToMapOfRootsToNotesOfKey: Map<Key, Map<Note, Note[]>> = ChordVoicingGenerator.computeMapOfKeyToMapOfRootsToNotesOfKey(
            searchedKeys,
            allowedKeyRoots
        );

        const filters: ((voicing: ChordVoicing) => boolean)[] = [
            (voicing) => voicing.smallestDistanceBetweenVoices() >= minimumSemitonesBetweenNotesOfTheVoicing,
            (voicing) => voicing.isCompatibleWithAnyOfTheseKeys(keyToMapOfRootsToNotesOfKey),
            (voicing) => voicing.getRepresentedChord()!.getNotes().size <= maxNumberOfDifferentNotes,
            (voicing) => voicing.getRepresentedChord()!.getNotes().size >= minNumberOfDifferentNotes,
            (voicing) =>
                !forbidTheSameNoteOnTheSameOctaveOnDifferentStrings ||
                !voicing.getHasSeveralTimesTheSameNoteOnTheSameOctave(),
            (voicing) => {
                let numberOfOpenStrings = 0;
                for (const entry of voicing.getFrettings().entries()) {
                    if (entry[1].isOpen) {
                        numberOfOpenStrings++;
                    }
                }
                return numberOfOpenStrings >= mustContainAtLeastThisManyOpenStrings;
            },
            (voicing) => {
                let numberOfOpenStrings = 0;
                for (const entry of voicing.getFrettings().entries()) {
                    if (entry[1].isOpen) {
                        numberOfOpenStrings++;
                    }
                }
                return numberOfOpenStrings <= mustContainAtMostThisManyOpenStrings;
            },
            (voicing) => {
                let passed = true;
                for (let stringNumber = 0; stringNumber < minimumNumberOfMutedLowestStrings; stringNumber++) {
                    passed = passed && FretAction.MUTE.equals(voicing.getFrettings().get(stringNumber));
                }
                return passed;
            },
            (voicing) => {
                let passed = true;
                for (
                    let stringNumber = tuning.getNumberOfStrings() - 1;
                    stringNumber + minimumNumberOfMutedHighestStrings >= tuning.getNumberOfStrings();
                    stringNumber--
                ) {
                    passed = passed && FretAction.MUTE.equals(voicing.getFrettings().get(stringNumber));
                }
                return passed;
            },
            (voicing) => {
                return highestNoteAllowed === null || voicing.getHighestNote(tuning)!.isLowerThan(highestNoteAllowed);
            },
            (voicing) => {
                return lowestNoteAllowed === null || voicing.getLowestNote(tuning)!.isHigherThan(lowestNoteAllowed);
            },
            (voicing) => {
                return !hideChordVoicingsThatDoNotHaveAThirdOrAFifth || voicing.doAnyOfTheRepresentationsHaveAThirdOrAFifth(tuning);
            },
        ];

        const filePath = './Voicings.txt';
        console.log(`Printing to ${filePath}`);
        const targetFile: PrintStream = new PrintStream(filePath);

        try {
            ChordVoicingGenerator.generateAllVoicings(
                tuning,
                fretSpan,
                firstFretToScan,
                lastFretToScan,
                filters,
                targetFile,
                keyToMapOfRootsToNotesOfKey,
                hideChordVoicingsThatDoNotHaveAThirdOrAFifth,
                hideChordVoicingsWithb5,
                hideChordVoicingsWithsharp5,
                hideChordVoicingsWithb7,
                hideChordVoicingsWithAdd7,
                hideChordVoicingsWithb9,
                hideChordVoicingsWithAdd9,
                hideChordVoicingsWithb11,
                hideChordVoicingsWithAdd11,
                hideChordVoicingsWithb13,
                hideChordVoicingsWithAdd13
            );
        } finally {
            targetFile.closeAndDownload();
        }

        console.log('Stream closed');
    }

    private static computeMapOfKeyToMapOfRootsToNotesOfKey(
        searchedKeys: Key[],
        allowedRoots: Note[]
    ): Map<Key, Map<Note, Note[]>> {
        const keyToMapOfRootsToNotesOfKey = new Map<Key, Map<Note, Note[]>>();

        console.log('in computeMapOfKeyToMapOfRootsToNotesOfKey');

        for (const searchedKey of searchedKeys) {
            const rootsToNotesOfKey = new Map<Note, Note[]>();

            for (const root of Object.values(Note)) {
                if (
                    allowedRoots === null ||
                    allowedRoots.length === 0 ||
                    allowedRoots.includes(root)
                ) {
                    const notesInKeyStartingOnRoot: Note[] = [];
                    const semitonesFromRootList = searchedKey.getSemitonesFromRootList();

                    for (const semitonesFromRoot of semitonesFromRootList) {
                        notesInKeyStartingOnRoot.push(root.up(semitonesFromRoot));
                    }

                    rootsToNotesOfKey.set(root, notesInKeyStartingOnRoot);
                }
            }

            keyToMapOfRootsToNotesOfKey.set(searchedKey, rootsToNotesOfKey);
        }

        return keyToMapOfRootsToNotesOfKey;
    }

    private static generateAllVoicings(
        tuning: Tuning,
        fretSpan: number,
        firstFretToScan: number,
        lastFretToScan: number,
        filters: ((voicing: ChordVoicing) => boolean)[],
        output: PrintStream,
        keyToMapOfRootsToNotesOfKey: Map<Key, Map<Note, Note[]>>,
        hideChordVoicingsThatDoNotHaveAThirdOrAFifth: boolean,
        hideChordVoicingsWithb5: boolean,
        hideChordVoicingsWithsharp5: boolean,
        hideChordVoicingsWithb7: boolean,
        hideChordVoicingsWithAdd7: boolean,
        hideChordVoicingsWithb9: boolean,
        hideChordVoicingsWithAdd9: boolean,
        hideChordVoicingsWithb11: boolean,
        hideChordVoicingsWithAdd11: boolean,
        hideChordVoicingsWithb13: boolean,
        hideChordVoicingsWithAdd13: boolean
    ): void {
        console.log(`Generating voicings for the tuning ${tuning}`);

        const startEnds: StartAndEnd[] = [];

        let fretStart = firstFretToScan;
        let fretEnd: number;
        while ((fretEnd = fretStart + fretSpan - 1) <= lastFretToScan) {
            startEnds.push(new StartAndEnd(fretStart, fretEnd));
            fretStart++;
        }

        startEnds.forEach((startEnd) => {
            const voicings = this.generateVoicings(
                tuning,
                startEnd.start,
                startEnd.end,
                filters,
                keyToMapOfRootsToNotesOfKey
            );

            console.log(
                `Voicings between frets ${startEnd.start} and ${startEnd.end} generated successfully`
            );

            voicings.forEach((voicing) => {

                const fullRepresentations: string[] = voicing
                    .fullRepresentations(tuning,
                        hideChordVoicingsThatDoNotHaveAThirdOrAFifth,
                        hideChordVoicingsWithb5,
                        hideChordVoicingsWithsharp5,
                        hideChordVoicingsWithb7,
                        hideChordVoicingsWithAdd7,
                        hideChordVoicingsWithb9,
                        hideChordVoicingsWithAdd9,
                        hideChordVoicingsWithb11,
                        hideChordVoicingsWithAdd11,
                        hideChordVoicingsWithb13,
                        hideChordVoicingsWithAdd13);

                if (fullRepresentations.length != 0) {
                    output.println(`____________ ${voicing.noteRepresentation(tuning)}  `);

                    fullRepresentations.forEach((fullRepresentation) => {
                        output.println(
                            `_________ ${fullRepresentation}:  `
                        );
                    });

                    output.println(`${voicing.toString()}  `);
                    output.println(`${voicing.notesOnStrings()}  `);

                    const sj: string[] = [];
                    voicing.getCompatibleKeysAndRoots().forEach((compatibleKeyAndRoot) => {
                        sj.push(`${compatibleKeyAndRoot.root} ${compatibleKeyAndRoot.key}`);
                    });

                    output.println(`Diatonic to ${sj.join(', ')}  `);
                    output.println("__________________________  ");
                }
            });
        });

        output.println("All voicings generated successfully");
    }

    static generateVoicings(
        tuning: Tuning,
        lowestFretToScan: number,
        highestFretToScan: number,
        filters: ((voicing: ChordVoicing) => boolean)[],
        keyToMapOfRootsToNotesOfKey: Map<Key, Map<Note, Note[]>>
    ): Set<ChordVoicing> {

        console.log("in generateVoicings")

        const stringNumberToFretNumberToNote: Map<number, Map<number, NoteAndOctave>> = new Map();
        for (const [stringNumber, stringTuningNote] of tuning.getStringNotes()) {
            if (!stringNumberToFretNumberToNote.has(stringNumber)) {
                stringNumberToFretNumberToNote.set(stringNumber, new Map());
            }
            const fretNumberToNote = stringNumberToFretNumberToNote.get(stringNumber);

            if (fretNumberToNote) {
                for (let currentFret = lowestFretToScan; currentFret < highestFretToScan; currentFret++) {
                    const currentNote = stringTuningNote.up(currentFret);
                    fretNumberToNote.set(currentFret, currentNote);
                }
            }
        }

        const possibleFretActions: FretAction[] = [
            FretAction.MUTE,
            FretAction.OPEN,
            ...Array.from({ length: highestFretToScan - lowestFretToScan + 1 }, (_, i) => FretAction.hold(lowestFretToScan + i)),
        ];

        return this.generateVoicingsFromFretActions(possibleFretActions, tuning, filters, keyToMapOfRootsToNotesOfKey);
    }

    static generateVoicingsFromFretActions(
        possibleFretActions: FretAction[],
        tuning: Tuning,
        filters: ((voicing: ChordVoicing) => boolean)[],
        keyToMapOfRootsToNotesOfKey: Map<Key, Map<Note, Note[]>>
    ): Set<ChordVoicing> {

        console.log('in generateVoicingsFromFretActions');

        return this.iterateFrets(
            possibleFretActions,
            tuning,
            0,
            [],
            filters,
            keyToMapOfRootsToNotesOfKey
        );
    }

    static iterateFrets(
        possibleFretActions: FretAction[],
        tuning: Tuning,
        currentString: number,
        previousIterations: FretAction[],
        filters: ((voicing: ChordVoicing) => boolean)[],
        keyToMapOfRootsToNotesOfKey: Map<Key, Map<Note, Note[]>>
    ): Set<ChordVoicing> {

        const voicingsConstructed = new Set<ChordVoicing>();

        const stringsToIterate = tuning.getNumberOfStrings();

        for (const fretAction of possibleFretActions) {
            const thisIteration = previousIterations.concat(fretAction);

            if (currentString < stringsToIterate - 1) {

                let newVoicingsConstructed = this.iterateFrets(
                    possibleFretActions,
                    tuning,
                    currentString + 1,
                    thisIteration,
                    filters,
                    keyToMapOfRootsToNotesOfKey
                )

                for (let newVoicingConstructed of newVoicingsConstructed) {
                    voicingsConstructed.add(newVoicingConstructed);
                }

            } else {
                const chordVoicing = new ChordVoicing(thisIteration);

                if (chordVoicing.numberOfNonMutedStrings() > 1) {
                    chordVoicing.postProcessing(tuning, keyToMapOfRootsToNotesOfKey);
                    let passedAllFilters = true;

                    for (const filter of filters) {
                        if (!filter(chordVoicing)) {
                            passedAllFilters = false;
                            break;
                        }
                    }

                    if (passedAllFilters) {
                        voicingsConstructed.add(chordVoicing);
                    }
                }
            }
        }

        return voicingsConstructed;
    }

}